x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,309 @@
+				clientAddr = addr
+			}
+			clientAddrMu.Unlock()
+
+			_, err = upstreamConn.Write(buf[:n])
+			if err != nil {
+				errCh <- fmt.Errorf("error writing to upstream UDP: %v", err)
+				return
+			}
+		}
+	}()
+
+	// Upstream to client
+	go func() {
+		buf := make([]byte, 64*1024)
+		for {
+			n, err := upstreamConn.Read(buf)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading from upstream UDP: %v", err)
+				return
+			}
+
+			clientAddrMu.Lock()
+			if clientAddr != nil {
+				_, err = clientConn.WriteToUDP(buf[:n], clientAddr)
+				if err != nil {
+					errCh <- fmt.Errorf("error writing to client UDP: %v", err)
+					return
+				}
+			}
+			clientAddrMu.Unlock()
+		}
+	}()
+}
+func convertToNetAddr(ip netip.Addr, port uint16) *net.UDPAddr {
+	return &net.UDPAddr{
+		IP:   ip.AsSlice(),
+		Port: int(port),
+	}
+}
+
+type Config struct {
+	Server struct {
+		TCPListenAddress string `mapstructure:"tcp_listen_address"`
+		UDPListenAddress string `mapstructure:"udp_listen_address"`
+	}
+	SOAX struct {
+		PackageID  string `mapstructure:"package_id"`
+		PackageKey string `mapstructure:"package_key"`
+		Address    string `mapstructure:"address"`
+	}
+	Credentials map[string]string `mapstructure:"credentials"`
+}
+
+func loadConfig() (*Config, error) {
+	v := viper.New()
+	v.SetConfigName("config")
+	v.SetConfigType("yaml")
+	v.AddConfigPath(".")
+
+	if err := v.ReadInConfig(); err != nil {
+		return nil, fmt.Errorf("failed to read config file: %w", err)
+	}
+
+	var config Config
+	if err := v.Unmarshal(&config); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config: %w", err)
+	}
+
+	config.SOAX.PackageID = "package-" + config.SOAX.PackageID + "-"
+
+	if config.SOAX.Address == "" {
+		config.SOAX.Address = "proxy.soax.com:5000"
+	}
+
+	return &config, nil
+}
+
+func main() {
+
+	config, err := loadConfig()
+	if err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
+
+	// Create the custom authenticator
+	customAuth := CustomAuthenticator{
+		credentials: config.Credentials,
+	}
+
+	streamEndpoint := transport.StreamDialerEndpoint{Dialer: &transport.TCPDialer{}, Address: config.SOAX.Address}
+	client, err := csocks5.NewClient(&streamEndpoint)
+	if err != nil {
+		return
+	}
+
+	server := socks5.NewServer(
+		socks5.WithAuthMethods([]socks5.Authenticator{customAuth}),
+		socks5.WithDialAndRequest(func(ctx context.Context, network, addr string, req *socks5.Request) (net.Conn, error) {
+			authContext := req.AuthContext
+			if authContext == nil {
+				return nil, errors.New("no auth context available")
+			}
+
+			suffix, ok := authContext.Payload["suffix"]
+			if !ok {
+				return nil, errors.New("no suffix found in auth context")
+			}
+
+			// Construct the upstream username by concatenating the base username with the suffix
+			soaxUsername := config.SOAX.PackageID + suffix // e.g., "package-123456-country-ir-seesionid-..."
+			err = client.SetCredentials([]byte(soaxUsername), []byte(config.SOAX.PackageKey))
```

Comment:
This is a race condition, since it will change the credential for everyone using the client